### ISSUE TITLE
FR: nested network filter chains

### ISSUE THREAD

**Issue by kyessenov** [envoyproxy/envoy#18035]
Currently, listener inspectors and transport sockets are forced to reside at one-level: first inspect, then match, then use a transport socket. This is quite inflexible, and leads to situations where a whole expensive listener is needed (e.g. https://github.com/envoyproxy/envoy/issues/4076). The proposal is to add a oneof that allows (inspectors + filter chains) to reside within the filter chains next to filters. The semantics is that processing is staged where once a filter chain is selected, transport socket is applied, and then inspector + filter chain matching restart.

This solves a bunch of issues:
1. Proxy protocol can be placed within TLS on server-side.
2. TLS-within-TLS can be matched provided outer TLS is prior knowledge.
3. HTTP-within-TLS can be sniffed provided outer TLS is prior knowledge.

**Comment by mattklein123** [comment]
@kyessenov I'm having trouble quickly understanding the API shape of the proposal. Can you do a quick API mock up to share and discuss? Thank you.

**Comment by kyessenov** [comment]
@mattklein123 Opened PR https://github.com/envoyproxy/envoy/pull/18079 with a minimal change.

**Comment by mattklein123** [comment]
Thanks, that makes sense to me at a high level.

**Comment by soulxu** [comment]
@kyessenov let me know if you need any volunteer :)

**Comment by lizan** [comment]
Yeah at high level that makes sense to me. It also reminds whether we want FCDS / filter chain references to avoid duplicated filter chain configs.

**Comment by kyessenov** [comment]
This is probably still a useful feature, but we can do something better with matcher API and aligned transport / regular sockets.

**Comment by kyessenov** [comment]
Some additional thoughts that we had in this area:
* delaying inspectors until middle of the decision tree in the unified matchers: helps with reducing duplication with per-listener filter conditions
* this feels a lot like internal redirect in HTTP; perhaps, a better way to frame it is as a re-evaluation of the matcher tree.
